webapp/src/js/controllers/reports.js

```diff
@@ -329,24 +331,58 @@ angular
     $scope.$on('VerifyReport', function(e, valid) {
       if ($scope.selected[0].doc.form) {
         $scope.setLoadingSubActionBar(true);
+        const doc = $scope.selected[0].doc;
+
+        Auth('can_edit_verification')
+          .then(() => true)
+          .catch(() => false)
+          .then(function (canEditVerification) {
+            if (canEditVerification) {
+              return false;
+            }
+            
+            const docHasExistingResult = doc.verified !== undefined;
+            if (docHasExistingResult) {
+              return true;
+            }
 
-        var doc = $scope.selected[0].doc;
-        if (doc.contact) {
-          doc.contact = lineage.minifyLineage(doc.contact);
-        }
+            return Modal({
+              templateUrl: 'templates/modals/verify_confirm.html',
+              controller: 'VerifyReportModalCtrl',
+              model: {
+                proposedVerificationState: $translate.instant(valid ? 'reports.verify.valid' : 'reports.verify.invalid'),
+              },
+            });
```

Comment:
This section is quite hard to read. I can't think of a way to write the code any clearer, so I think it needs some comments above each return to explain what's going on.

Reply:
Ya... I never know what to do with this. I miss async!  

Attempting something. Let me know if you like it better.